### ISSUE TITLE
asyncio and gevent

### ISSUE THREAD

**Issue by 2mf** [gevent/gevent#982]
Would it be feasible to integrate an ayncio policy into gevent so that asyncio can run on gevent e.g. https://bitbucket.org/haypo/aiogevent and the other way around running gevent on asyncio e.g. https://github.com/decentfox/tulipcore

**Comment by jamadden** [comment]
If people have written the code, than apparently it is possible. But if there are existing projects that do this, I don't see a need for gevent to copy them at this time.

**Comment by 2mf** [comment]
Well I thought that could be integrated into gevent so changes in gevent can be more easily reflected onto those packages and as well for maintenance reasons, aiogevent and tulipcore (@fantix) were last updated in 2015 ...

So for my point of view that is a reasonable question.

**Comment by fantix** [comment]
Apologies 🤕

**Comment by 2mf** [comment]
In #900 asked for a similar thing

**Comment by arcivanov** [comment]
aiogevent is apparently gone.

**Comment by chadrik** [comment]
aiogevent lives on!   Luckily, someone made [a clone of it](https://github.com/2mf/aiogevent) before it was deleted.  It looked promising so I tracked down the original author (none other than the creator of [trollius](https://github.com/vstinner/trollius), Victor Stinner) to ask him if there was a particular reason that he abandoned the project.  He said, paraphrasing, "no, just lost interest". 

Using it is as simple as:

```python
asyncio.set_event_loop_policy(aiogevent.EventLoopPolicy())
```

I ran the asyncio tests from [autobahn](https://github.com/crossbario/autobahn-python/) patched to run using the aiogevent gevent-based event loop, and they all passed.  This is a reasonable demonstration that the eventloop works, but it doesn't prove that asyncio and gevent can peacefully coexist.  For that I've been working on a separate set of tests (more like demos): gist [here](https://gist.github.com/chadrik/a7b7d8abce891978a7dbfd0d086828b3).  Long story short, as long as your gevent-based code runs below a greenlet that has been specially prepared and handed to the ascyncio eventloop it should run cooperatively with asyncio. 

I'll report back with more information, but since this seems to be a hot topic, I thought I'd get the news out as soon as possible.

I'm not sure of @2mf relationship to the project, but there haven't been any updates in 3 years, and the original author does not want to be involved anymore, so it looks like the project is orphaned right now.  I would love to see this become part of gevent.  On the plus side, it has a pretty small footprint, comes with tests, and already seems to work quite well.

**Comment by arcivanov** [comment]
I would agree here. As asyncio is now a part of Python core, it makes sense to expose Gevent via asyncio.

**Comment by mahmoudimus** [comment]
@arcivanov and @chadrik, I'm happy to help out here. Exposing gevent via asyncio and vice versa will help bridge the gap and minimize the spread of `async` and `await` definitions everywhere.

**Comment by chadrik** [comment]
@mahmoudimus  I think the main thing that we need right now is to test `aiogevent`, and then if it appears to be the best solution for bridging the two async words, lobby to have it moved into the `gevent` project.

**Comment by aathan** [comment]
Any news/recent experiences with mixing asyncio and gevent?

**Comment by chadrik** [comment]
I'm no longer involved in this, but it would be cool if it were solved!

**Comment by mguijarr** [comment]
The support of asyncio is becoming more and more a problem, with the proliferation of asyncio libraries out there.

For example I would like full gevent support with [ptpython](https://github.com/prompt-toolkit/ptpython) ; the recent versions of the underlying [prompt-toolkit](https://github.com/prompt-toolkit/python-prompt-toolkit) rely on asyncio. I never want to have my gevent code "blocked" because of asyncio processing.

I am using aiogevent for this, so far so good - I guess it is a good, but it is scary to rely on unmaintained code.

It is only 300+ lines of code today... The main difficulty I guess is to implement `call_soon_threadsafe`. I do not know how to make progress on this.

**Comment by mguijarr** [comment]
@jamadden Hi Jason... Sorry to hail to you like this - I sent you a mail (to jason@nextthought.com) about this one, maybe it got lost somewhere. I send you my best wishes for 2021.

Another reason to have gevent + asyncio interoperability... Jupyer notebooks. It is possible to run a Jupyter kernel, that is able to run the gevent-based code by adding a kernel configured like this:

```
{
    "display_name": "Python w/gevent",
    "language": "python",
    "argv": ["python", "-c",
             "import gevent; from gevent.monkey import patch_all; patch_all(thread=False); import aiogevent; import asyncio;asyncio.set_event_loop_policy(aiogevent.EventLoopPolicy()); from ipykernel.kernelapp import main; gevent.spawn(main).get()",
             "-f",
             "{connection_file}"]
}
```

(new kernel as to be added to "kernelspec")

**Comment by FelixSchwarz** [comment]
Just wanted to mention that there is also a new project named [asyncio-gevent](https://github.com/gfmio/asyncio-gevent) which tackles the same problem (though the example to run asyncio on gevent [did not work for me](https://github.com/gfmio/asyncio-gevent/issues/1)).

**Comment by gfmio** [comment]
Just to chime in here, I fixed what I hope are all the bugs in [asyncio-gevent](https://github.com/gfmio/asyncio-gevent)’s asyncio-on-gevent runtime and updated the documentation in version `0.2.0`.

It actually turned out to be simpler to implement than I had initially assumed, because the you can use the gevent selector with the default asyncio selector-based event loop and it seems to work as expected.

However, I’m also still working on implementing the reverse: Running gevent on top of asyncio by implementing a gevent `ILoop` which internally uses asyncio. Contributions welcome! :)

Example:

```py3
import gevent.monkey
gevent.monkey.patch_all()

import asyncio

import asyncio_gevent

asyncio.set_event_loop_policy(asyncio_gevent.EventLoopPolicy())

import gevent

def blocking_function() -> int:
    gevent.sleep(10)
    return 42


async def main() -> None:
    greenlet = gevent.spawn(blocking_function)
    future = asyncio_gevent.greenlet_to_future(greenlet)
    result = await future

asyncio.run(main())
```

**Comment by christophehenry** [comment]
@gfmio Should asyncio-gevent be considered unmaintained? There are a bunch of unanswered issues a one unmerged PR.
